app/models/user.rb

```diff
@@ -0,0 +1,19 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+
+  validates :name, presence: true,
+            length: {maximum: Settings.user.NAME_MAX_LENGTH}
+  validates :email, presence: true,
+            length: {minimum: Settings.user.EMAIL_MIN_LENGTH,
+                     maximum: Settings.user.EMAIL_MAX_LENGTH},
+            format: {with: Settings.user.VALID_EMAIL_REGEX},
+            uniqueness: {case_sensitive: false}
+  validates :password, presence: true,
+            length: {minimum: Settings.user.PASSWORD_MIN_LENGTH}, if: :password
+  has_secure_password
```

Comment:
dòng này làm gì em?

Reply:
"has_secure_password" là một hàm thuộc Gem "bcrypt". Hàm này có tác dụng xác nhận password bằng cách sinh ra hai trường ảo (không lưu vào cơ sở dữ liệu) "password" và "password_confirmation" cho mỗi bản ghi. Chỉ khi giá trị của hai trường này bằng nhau thì bản ghi đó mới được lưu vào cơ sở dữ liệu.

"has_secure_password" mã hóa mật khẩu trước khi lưu vào cơ sở dữ liệu, đảm bảo an toàn thông tin cho người dùng. Mật khẩu sau khi mã hóa được lưu trong trường password_digest.

"has_secure_password" hỗ trợ việc xác thực tài khoản của người dùng bằng mật khẩu (authenticate).